Add term entry kotlin datetime method todayin 3866

aa-arroyo opened the pull request:
<!--
👋 Hi, thanks for submitting a PR to Codecademy Docs! 💖
Please fill out all fields below to ensure your PR is reviewed quickly.

**IMPORTANT**

If you would like to receive credit for your contribution to an entry, make sure to have your Codecademy user profile linked to your GitHub account:

1. Go to your Codecademy dashboard.
2. Click on your profile image in the top-right and then choose "Profile".
3. Then click "Edit Profile".
4. In the "GitHub Username" field, add your username (without the @).
5. Click "Save Changes"

Of course, you can opt not to do this and be listed as an "Anonymous contributor", instead. :)
-->

### Description

<!-- Please write a summary of the change, such as which topic(s) and file(s) that you have edited or created. Please also include relevant motivation and context: -->

### Type of Change

<!--- Please delete or cross off the bullet point(s) that are irrelevant to this PR: -->

- Adding a new entry: I am adding a term entry (#3866). Kotlin datetime, todayIn() method. I have created a directory named todayIn and a todayIn.md file. I am working on this entry as part of the Learn Git & GitHub course at Codecademy. 
This is my first time learning about Kotlin/Kotlinx. Setting up the IntelliJ IDEA IDE and adding the kotlinx.datetime dependency took me some time, but I eventually got it to work and could play around with the code and better understand it. Any guidance and feedback will be much appreciated.

Closes #3866 

### Checklist

<!-- Please check ALL the boxes: -->

- [Y] All writings are my own.
- [Y] My entry follows the Codecademy Docs style guide.
- [Y] My changes generate no new warnings.
- [Y] I have performed a self-review of my own writing and code.
- [Y] I have checked my entry and corrected any misspellings.
- [Y] I have made corresponding changes to the documentation if needed.
- [Y] I have confirmed my changes are not being pushed from my forked `main` branch.
- [Y] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [Y] Under "Development" on the right, I have linked any issues that are relevant to this PR (write "Closes #<issue number> in the "Description" above).

<!--
Having trouble with the PR checker? Here are some common issues and resolutions:

- verify_formatting is failing
  - run `yarn format path/to/markdown/file.md` or `yarn format:all` and commit the results
- verify_lint is failing
  - same as above
  - if verify_lint is still failing, running `yarn lint` locally should let you know what needs to be changed by hand
- test is failing
  - ensure any new markdown files have a `Title` and `Description` defined in their metadata
  - ensure any new markdown files only contain alphanumerics and dashes in their file names and have the same name as their parent directory
  - if that looks ok, running `yarn test` locally should let you know what the issue is
-->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Codecademy/docs?pullRequest=3938) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Codecademy/docs?pullRequest=3938) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Codecademy/docs?pullRequest=3938) it.</sub>

avdhoottt commented on content/kotlin/concepts/datetime/terms/todayIn/todayIn.md:
```diff
@@ -0,0 +1,38 @@
+---
+Title: '.todayIn()' # Required; the file name should be the same as the title, but lowercase, with dashes instead of spaces, and all punctuation removed
+Description: 'Passing a specific time zone as an argument and calling this method on an instant value of Clock will return the current date without time.' # Required; ideally under 150 characters and starts with a present-tense verb (used in search engine results and content previews)
+Subjects: # Please only use Subjects in the subjects.md file (https://github.com/Codecademy/docs/blob/main/documentation/subjects.md). If that list feels insufficient, feel free to create a new Subject and add it to subjects.md in your PR!
+  - 'Computer Science'
+  - 'Code Foundations'
+Tags: # Please only use Tags in the tags.md file (https://github.com/Codecademy/docs/blob/main/documentation/tags.md). If that list feels insufficient, feel free to create a new Tag and add it to tags.md in your PR!
+  - 'Methods'
+  - 'Date'
+  - 'Kotlin'
+CatalogContent: # Please use course/path landing page slugs, rather than linking to individual content items. If listing multiple items, please put the most relevant one first
+  - 'learn-kotlin'
+  - 'paths/computer-science'
+---
```
Remove the comments from the metadata as those comments are only for your convenience.

aa-arroyo replied:
Fixed.

avdhoottt commented on content/kotlin/concepts/datetime/terms/todayIn/todayIn.md:
```diff
@@ -0,0 +1,38 @@
+
+## Syntax
+
+
```
```suggestion
```
Remove the extra space between the lines.

aa-arroyo replied:
Fixed.

avdhoottt commented on content/kotlin/concepts/datetime/terms/todayIn/todayIn.md:
````diff
@@ -0,0 +1,38 @@
+## Syntax
+
+
+```
+Clock.todayIn(timeZone: TimeZone)
+```
````
Please go through the [style guide](https://github.com/Codecademy/docs/blob/main/documentation/style-guide.md) and format the code blocks and syntax as mentioned.

aa-arroyo replied:
I reviewed the style guide and made the necessary corrections to the todayIn.md file.

avdhoottt commented on content/kotlin/concepts/datetime/terms/todayIn/todayIn.md:
```diff
@@ -0,0 +1,38 @@
+The `.todayIn()` method will use an instant value from `Clock` and a time `TimeZone` argument to return a date in a YYYY-MM-DD format.
+## Example
+
+
```
```suggestion
```

aa-arroyo replied:
Fixed.

avdhoottt commented on content/kotlin/concepts/datetime/terms/todayIn/todayIn.md:
````diff
@@ -0,0 +1,38 @@
+## Example
+
+
+```
+fun main() {
+    // Clock.System and TimeZone.currentSystemDefault() queries the operating system time and time zone, respectively.
+    val myVar = Clock.System.todayIn(timeZone = TimeZone.currentSystemDefault())
+
+    println(myVar)
+    // Output: YYYY-MM-DD
+
+}
+```
````
The same goes for this code, go through the style guide and format this code block correctly. Display the output in a different code block, you can use the shell for it.

aa-arroyo replied:
Fixed.

avdhoottt reviewed (changes requested):
@aa-arroyo First of all, my suggestion would be that go through other entries of the docs, and observe how the content and code blocks are formatted. Take reference from the style guide and make the changes in this file.

aa-arroyo commented:
avdhoottt, thank you for the detailed feedback. I will work on all the requested changes.

aa-arroyo commented:
@avdhoottt I have corrected the todayIn.md file according to the feedback provided and the style guide. I added, committed, and pushed my changes. Please let me know if any additional actions are required from me, and I will get it done as soon as possible.

Thank you.

avdhoottt reviewed (changes requested):
@aa-arroyo I appreciate your work but I said to remove the line spaces from the code only and keep the other required spaces. For example, add a line space before and after the syntax and example code blocks. Please add the required spaces and it's good to go for the second review.

aa-arroyo commented:
@avdhoottt Added the required spacing.

avdhoottt commented on content/kotlin/concepts/datetime/terms/todayIn/todayIn.md:
````diff
@@ -0,0 +1,45 @@
+
+```kotlin
+// Import all classes from kotlinx.datetime package.
+import kotlinx.datetime.* 
+
+fun main() {
+  val myLocalDate = Clock.System.todayIn(timeZone = TimeZone.currentSystemDefault())
+
+  println(myLocalDate)
+}
````
```suggestion
import kotlinx.datetime.* 

fun main() {
  val myLocalDate = Clock.System.todayIn(timeZone = TimeZone.currentSystemDefault())
  println(myLocalDate)
}
```

aa-arroyo replied:
Removed blank line 41.

avdhoottt commented on content/kotlin/concepts/datetime/terms/todayIn/todayIn.md:
```diff
@@ -0,0 +1,44 @@
+
+The **`.todayIn()`** method returns a date when called on an Instant value of `Clock` while passing a `TimeZone` argument.
+
+## Syntaxx
```
```suggestion
## Syntax
```

avdhoottt commented on content/kotlin/concepts/datetime/terms/todayIn/todayIn.md:
```diff
@@ -0,0 +1,44 @@
+
+- `Clock.System`: records a moment in time, also known as an Instant (e.g., 2023-11-21, 00:00:00 UTC+0). Two possible ways to get an Instant are by using `Clock.now` or `Clock.System`.
+- `TimeZone.currentSystemDefault()`: pulls and records an instance of the current system time zone (e.g., America/Los_Angeles).
+- The return value is a date in the following format YYYY-MM-DD (e.g., 2023-11-21)
```
```suggestion
```

avdhoottt reviewed (approved):
Looks Good for the second review.

SSwiniarski commented on content/kotlin/concepts/datetime/terms/todayIn/todayIn.md:
```diff
@@ -0,0 +1,43 @@
+---
+Title: '.todayIn()'
+Description: 'Passing a specific time zone as an argument and calling this method on an Instant value of Clock will return a date.'
```
```suggestion
Description: 'Returns a date when calling this method on a Clock with a specific time zone as an argument.'
```

SSwiniarski commented on content/kotlin/concepts/datetime/terms/todayIn/todayIn.md:
```diff
@@ -0,0 +1,43 @@
+  - 'paths/computer-science'
+---
+
+The **`.todayIn()`** method returns a date when called on an Instant value of `Clock` while passing a `TimeZone` argument.
```
```suggestion
The **`.todayIn()`** method returns a date when called on a `Clock` while passing a `TimeZone` argument.
```
A clock isn't an instant value, it is a source of instant values.

SSwiniarski commented on content/kotlin/concepts/datetime/terms/todayIn/todayIn.md:
````diff
@@ -0,0 +1,43 @@
+## Syntax
+
+```pseudo
+Clock.System.todayIn(timeZone: TimeZone.currentSystemDefault())
````
```suggestion
myClock.todayIn(myTimeZone)
```
The Syntax example needs to be more generic.  (It can be any Clock and any TimeZone)

SSwiniarski commented on content/kotlin/concepts/datetime/terms/todayIn/todayIn.md:
````diff
@@ -0,0 +1,43 @@
+Clock.System.todayIn(timeZone: TimeZone.currentSystemDefault())
+```
+
+- `Clock.System`: records a moment in time, also known as an Instant (e.g., 2023-11-21, 00:00:00 UTC+0). Two possible ways to get an Instant are by using `Clock.now` or `Clock.System`.
````
```suggestion
- `myClock`: is a Clock providing a source for Instants (e.g., 2023-11-21, 00:00:00 UTC+0). `Clock.System` is the Clock that refers to the current platform for the current time.
```

SSwiniarski commented on content/kotlin/concepts/datetime/terms/todayIn/todayIn.md:
````diff
@@ -0,0 +1,43 @@
+```
+
+- `Clock.System`: records a moment in time, also known as an Instant (e.g., 2023-11-21, 00:00:00 UTC+0). Two possible ways to get an Instant are by using `Clock.now` or `Clock.System`.
+- `TimeZone.currentSystemDefault()`: pulls and records an instance of the current system time zone (e.g., America/Los_Angeles).
````
```suggestion
- `myTimeZone`: An instance of the TimeZone class. (e.g., America/Los_Angeles).
```

SSwiniarski commented on content/kotlin/concepts/datetime/terms/todayIn/todayIn.md:
```diff
@@ -0,0 +1,43 @@
+- `TimeZone.currentSystemDefault()`: pulls and records an instance of the current system time zone (e.g., America/Los_Angeles).
+
+## Example
+
```
There should be an introductory sentence summarizing what the example is doing.

SSwiniarski commented on content/kotlin/concepts/datetime/terms/todayIn/todayIn.md:
````diff
@@ -0,0 +1,43 @@
+}
+```
+
+This example results in the following output:
````
```suggestion
This example results in something like the following output:
```
The code as written will not always return the same date.

SSwiniarski reviewed (changes requested):
@aa-arroyo I did a second review and had a number of changes to suggest. Especially to clean up the confusion between a Clock and an Instant, they're two different things.

aa-arroyo commented:
Hi @SSwiniarski 

I made all the suggested changes and added an introductory sentence summarizing the example. I tried pushing my changes like I did before, but I got this error:

To https://github.com/aa-arroyo/docs.git
 ! [rejected]          add-term-entry-kotlin-datetime-method-todayin-3866 -> add-term-entry-kotlin-datetime-method-todayin-3866 (non-fast-forward)
error: failed to push some refs to 'https://github.com/aa-arroyo/docs.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. If you want to integrate the remote changes,
hint: use 'git pull' before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

I tried `git pull origin main` and then tried pushing from the feature branch (as I have been doing until now), but it's still not working.

On my forked feature branch under the Sync Fork option, it reads as follows: This branch is not behind the upstream Codecademy:main
No new commits to fetch. Enjoy your day!

I followed the steps listed on the link below, but it did not fix the error.
https://docs.github.com/en/get-started/using-git/dealing-with-non-fast-forward-errors

I am a bit lost. Any help will be appreciated.

SSwiniarski commented:
@caupolicandiaz , @KTom101 any idea on how to resolve their problem?

caupolicandiaz commented:
> Hi @SSwiniarski
> 
> I made all the suggested changes and added an introductory sentence summarizing the example. I tried pushing my changes like I did before, but I got this error:
> 
> To https://github.com/aa-arroyo/docs.git ! [rejected] add-term-entry-kotlin-datetime-method-todayin-3866 -> add-term-entry-kotlin-datetime-method-todayin-3866 (non-fast-forward) error: failed to push some refs to 'https://github.com/aa-arroyo/docs.git' hint: Updates were rejected because the tip of your current branch is behind hint: its remote counterpart. If you want to integrate the remote changes, hint: use 'git pull' before pushing again. hint: See the 'Note about fast-forwards' in 'git push --help' for details.
> 
> I tried `git pull origin main` and then tried pushing from the feature branch (as I have been doing until now), but it's still not working.
> 
> On my forked feature branch under the Sync Fork option, it reads as follows: This branch is not behind the upstream Codecademy:main No new commits to fetch. Enjoy your day!
> 
> I followed the steps listed on the link below, but it did not fix the error. https://docs.github.com/en/get-started/using-git/dealing-with-non-fast-forward-errors
> 
> I am a bit lost. Any help will be appreciated.

I believe you want to do `git pull origin feature-branch-name`, not main. Then, you can commit the changes pulled down. After the changes are downloaded the window will switch to the vim editor, use character `I` to type your commit message. Hit escape and then `:wq` to exit. Hopefully, that resolves your situation.

aa-arroyo commented:
Hi @caupolicandiaz 

I tried the suggested command and followed the hints/instructions provided by git:

mypc todayIn % git pull origin add-term-entry-kotlin-datetime-method-todayin-3866         
From https://github.com/aa-arroyo/docs
 * branch              add-term-entry-kotlin-datetime-method-todayin-3866 -> FETCH_HEAD
hint: You have divergent branches and need to specify how to reconcile them.
hint: You can do so by running one of the following commands sometime before
hint: your next pull:
hint: 
hint:   git config pull.rebase false  # merge
hint:   git config pull.rebase true   # rebase
hint:   git config pull.ff only       # fast-forward only
hint: 
hint: You can replace "git config" with "git config --global" to set a default
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
hint: or --ff-only on the command line to override the configured default per
hint: invocation.
fatal: Need to specify how to reconcile divergent branches.


mypc todayIn % git pull --ff-only origin add-term-entry-kotlin-datetime-method-todayin-3866
From https://github.com/aa-arroyo/docs
 * branch              add-term-entry-kotlin-datetime-method-todayin-3866 -> FETCH_HEAD
hint: Diverging branches can't be fast-forwarded, you need to either:
hint: 
hint: 	git merge --no-ff
hint: 
hint: or:
hint: 
hint: 	git rebase
hint: 
hint: Disable this message with "git config advice.diverging false"
fatal: Not possible to fast-forward, aborting.


mypc todayIn % git merge --no-ff origin add-term-entry-kotlin-datetime-method-todayin-3866
Already up to date.

mypc todayIn % git status
On branch add-term-entry-kotlin-datetime-method-todayin-3866
nothing to commit, working tree clean

There was no Vim window. I tried pushing again and got the same error as before.

aa-arroyo commented:
![feature-branch-history](https://github.com/Codecademy/docs/assets/71912479/31db6ab1-2f61-4005-b476-17fdd0f3836c)

I got even more confused looking at this graph on VScode... It seems there are two feature branches with the same name (?). It seems I messed up somewhere along the way, I just don't know exactly how.

caupolicandiaz commented:
@SSwiniarski I went ahead and pushed the changes. 
@aa-arroyo Unfortunately, it can be very difficult to troubleshoot an issue remotely. I would advise that you review our documentation for making and updating a PR. And try again on another entry.

SSwiniarski reviewed (approved):
@aa-arroyo Looks good.

aa-arroyo commented:
@SSwiniarski @caupolicandiaz I understand. Thank you for all the help along the way! Sorry, I couldn't produce the best entry version on the first attempt. 

Cheers.
